### PR TITLE
fix exception serialization

### PR DESCRIPTION
**Pull request by danielnorberg**
https://github.com/EsotericSoftware/kryo/issues/532

### CODE REVIEW

File: flo-freezer/src/test/java/com/spotify/flo/freezer/PersistingContextTest.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    final Exception exception = exception1;
+
+    final Path exceptionFile = temporaryFolder.getRoot().toPath().resolve("exception");
+    PersistingContext.serialize(exception, exceptionFile);
+
+    final Exception deserialized = PersistingContext.deserialize(exceptionFile);
+
+    // Verify that stack trace can be accessed and printed and seems to be correct
+
+    deserialized.printStackTrace();
```
Comment (honnix) [inline flo-freezer/src/test/java/com/spotify/flo/freezer/PersistingContextTest.java]: Is this needed?

File: flo-freezer/src/test/java/com/spotify/flo/freezer/PersistingContextTest.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    assertThat(stacktrace, containsString("Suppressed: com.spotify.flo.freezer.PersistingContextTest$FoobarException"));
+    assertThat(stacktrace, containsString("Caused by: java.lang.RuntimeException: bar"));
+    assertThat(stacktrace, containsString("Suppressed: java.io.IOException: baz"));
+    assertThat(stacktrace, containsString("Caused by: java.lang.InterruptedException: quux"));
+
+    assertThat(deserialized.getStackTrace().length, is(not(0)));
+    assertThat(deserialized.getStackTrace()[0].getClassName(), is("com.spotify.flo.freezer.PersistingContextTest"));
+    assertThat(deserialized.getStackTrace()[0].getMethodName(), is("exceptionSerialization"));
+
+    for (StackTraceElement stackTraceElement : deserialized.getStackTrace()) {
```
Comment (honnix) [inline flo-freezer/src/test/java/com/spotify/flo/freezer/PersistingContextTest.java]: Here as well.

File: flo-freezer/src/test/java/com/spotify/flo/freezer/PersistingContextTest.java (hunk 1 above)
Comment (danielnorberg) [reply flo-freezer/src/test/java/com/spotify/flo/freezer/PersistingContextTest.java]: Yes, like the comment says, we explicitly want to verify this.

The problem here was that exceptions that were deserialized would throw NPE when `printStackTrace()` was called on them

File: flo-freezer/src/test/java/com/spotify/flo/freezer/PersistingContextTest.java (hunk 2 above)
Comment (danielnorberg) [reply flo-freezer/src/test/java/com/spotify/flo/freezer/PersistingContextTest.java]: Yes.

File: flo-freezer/src/test/java/com/spotify/flo/freezer/PersistingContextTest.java (hunk 1 above)
Comment (honnix) [reply flo-freezer/src/test/java/com/spotify/flo/freezer/PersistingContextTest.java]: But the lines below uses print writer to verify each line explicitly.

File: flo-freezer/src/test/java/com/spotify/flo/freezer/PersistingContextTest.java (hunk 2 above)
Comment (honnix) [reply flo-freezer/src/test/java/com/spotify/flo/freezer/PersistingContextTest.java]: Just concerning about unnecessary noise in `System.err`, and it is not clear to me what we are testing here.

Comment (danielnorberg) [reply flo-freezer/src/test/java/com/spotify/flo/freezer/PersistingContextTest.java]: Ok I guess we can remove this loop.

File: flo-freezer/src/test/java/com/spotify/flo/freezer/PersistingContextTest.java (hunk 1 above)
Comment (danielnorberg) [reply flo-freezer/src/test/java/com/spotify/flo/freezer/PersistingContextTest.java]: Will remove the loop below.

File: flo-freezer/src/test/java/com/spotify/flo/freezer/PersistingContextTest.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    final Exception deserialized = PersistingContext.deserialize(exceptionFile);
+
+    // Verify that stack trace can be accessed and printed and seems to be correct
+
+    deserialized.printStackTrace();
+
+    final StringWriter sw = new StringWriter();
+    final PrintWriter pw = new PrintWriter(sw);
+    deserialized.printStackTrace(pw);
```
Comment (honnix) [inline flo-freezer/src/test/java/com/spotify/flo/freezer/PersistingContextTest.java]: I mean we do this anyway here so why printing to stdout at line 55?

**Comment by danielnorberg** [comment]
I’m paranoid

**Comment by honnix** [comment]
👍

**Pull request merged**